### ISSUE TITLE
fabric8 console not loading on kubernetes vagrant 1.0.2 release

### ISSUE THREAD

**Issue by rawlingsj** [fabric8io/fabric8#4646]
Running Kubernetes and installing the console using...

```
export VAGRANT_DEFAULT_PROVIDER=virtualbox
export KUBERNETES_PROVIDER=vagrant
curl -L https://github.com/kubernetes/kubernetes/releases/download/v1.0.2/kubernetes.tar.gz | tar -xvz
cd kubernetes
./cluster/kube-up.sh

kubectl create -f http://central.maven.org/maven2/io/fabric8/apps/console-kubernetes/2.2.19/console-kubernetes-2.2.19-kubernetes.json
```

Once the pods are running I can curl the console using 
`curl --user vagrant:vagrant https://10.245.1.2/api/v1/proxy/namespaces/default/services/fabric8/`

But not able to load in a browser (tried chrome and firefox).  The JS console does have a 404 and another error but not sure if that's a red herring as it did load once in a browser tab after I left my laptop and returned 15 mins later, but unable to get it to work since.

```
Uncaught Error: Cannot find module 'function (e){"object"==typeof exports&&"object"==typeof module?e(require("../../lib/codemirror"),require("../fold/xml-fold")):"function"==typeof define&&define.amd?define(["../../lib/codemirror","../fold/xml-fold"],e):e(CodeMirror)}'i @ libs-9a358b6e.js:152(anonymous function) @ libs-9a358b6e.js:178
10.245.1.2/:165 GET https://10.245.1.2/api/v1/proxy/namespaces/default/services/fonts/PatternFlyIcons-webfont.ttf 
10.245.1.2/:1 GET https://10.245.1.2/api/v1/proxy/namespaces/default/services/fonts/PatternFlyIcons-webfont.woff 404 (Not Found)
```

Seems to be the same symptoms as the kube ui is having https://github.com/kubernetes/kubernetes/issues/12692

**Comment by jimmidyson** [comment]
Notice the path doesn't include the fabric8 service name. Assuming that's the problem...

**Comment by rawlingsj** [comment]
Yeah you're right, the 404 when getting the font doesn't have the service name.  Having said that I just checked the browser tab that the console did load in and it also had the 404 font errors so, I think that's a separate issue to the reason the console wont load in at all apart from this one browser tab.

**Comment by davsclaus** [comment]
Its been a while, dont this work now?

**Comment by rawlingsj** [comment]
We're just waiting for the next Kubernetes release to test the fix.  Its been a month since the last release so hopefully soon.

**Comment by mukpong** [comment]
Guys any progress on this? I have deployed both fabric8 console 2.2.19 and 2.2.23.1 on google cloud GKE following the instructions on [fabric8 getting started] (http://fabric8.io/guide/getStarted/gke.html). I actually successfully deploy the fabric8 console. I have also deployed my own codes (built out of the war archetype quick starts) as pods on GKE.

The only problem is that I cannot really see the specific console of my deployed pod, which is where I would expect to see the camel routes etc. When you click on the link arrow from the pods pages (which opens a new browser to show you the specific console of the pod), it opens a blank page.

Checking the Javascript page of the blank page reveals this:

```
Uncaught Error: Cannot find module 'function (e){"object"==typeof exports&&"object"==typeof module?e(require("../../lib/codemirror"),require("../fold/xml-fold")):"function"==typeof define&&define.amd?define(["../../lib/codemirror","../fold/xml-fold"],e):e(CodeMirror)}'
libs-9a358b6e.js:11 [api-manager] Loaded the config.js config!
https://104.196.5.20/api/v1/proxy/namespaces/default/services/fonts/PatternFlyIcons-webfont.ttf Failed to load resource: the server responded with a status of 404 (Not Found)
```

How do I proceed from here?

**Comment by rawlingsj** [comment]
@michaelukpong sorry for the delay, a lot of the core team have been involved in an event recently.

@gashcrumb not sure if this error `https://104.196.5.20/api/v1/proxy/namespaces/default/services/fonts/PatternFlyIcons-webfont.ttf Failed to load resource: the server responded with a status of 404 (Not Found)` is the reason we cant connect to the pods but if I get a kubernetes env setup on GKE later would you mind taking a look to see if there's anything obvious?

**Comment by gashcrumb** [comment]
@rawlingsj Does that vagrant image work in vagrant/kubernetes in the fabric8-installer repo?  Can give that a try locally if so.  Doubt the issue is caused by failing to fetch those icons BTW, also the other mentioned issue with codemirror shouldn't cause a problem like this, think I've got it fixed in more recent versions of the console...

**Comment by michaelukpong** [comment]
@gashcrumb @rawlingsj  note that locally i have setup a vagrant/openshift (not kubernetes) based on the getting started guide. and I deployed both template pods (e.g. amqbroker) and my own pods. No issues with displaying the pod console.

I did try to `vagrant up` on the vagrant/kubernetes locally but somehow the box never started, **some errors popped up during the `vagrant up` process**. I didn't bother proceeding or trying to figure it out because for us our main focus is to deploy on GKE. 

Once i moved to GKE, I successfully deployed fabric8 console  `kc create -f ....json` and then deployed one of the template pods (amqbroker). Green status shows its successfully deployed. but you simply cannot navigate to its pod console --> You get a blank page and the Javascript console shows the error above.

**Comment by gashcrumb** [comment]
@michaelukpong ah, okay, then probably not worth testing against vagrant then.

@rawlingsj if you get a chance to set this up on GKE I'd be happy to take a peek.

**Comment by michaelukpong** [comment]
@gashcrumb  if it makes it faster you could peek at mine :) at this stage for us its still Proof of Concept to make a decision so no trade secrets in there :D 

Just a suggestion if it helps....  and yea I will go back and see if i can crack setting up the vagrant/kubernetes locally. I suspect i might get the same issue....

**Comment by gashcrumb** [comment]
@michaelukpong that'd work too :-) can you hit me up on IRC at #fabric8 on freenode?

**Comment by gashcrumb** [comment]
@michaelukpong btw will be afk for a bit, so if I take awhile to respond on IRC don't take it personally :-)

**Comment by michaelukpong** [comment]
@gashcrumb  haha, thats fine will do so in a few... in a longish meeting. Thanks mate

**Comment by gashcrumb** [comment]
@michaelukpong this looks a lot like this behavior -> https://bugzilla.redhat.com/show_bug.cgi?id=1243317

Basically the k8s master needs to be part of the SDN, as the console gets to jolokia via a little HTTP proxy in the API server.  If the API server can't connect to the internal IP of the pod then the connection will just timeout and fail.

**Comment by michaelukpong** [comment]
@gashcrumb and @rawlingsj thanks a lot so far. my apologies i lost connection on the IRC. Im actually doing this on my lappy while on a road trip :D anyways. so let me look at what you said in detail... thanks!

**Comment by michaelukpong** [comment]
@gashcrumb thanks.... I tried to be go solve it as you hinted, studied the link you sent in detail... studied kubernetes network.. but I'm sorry im stumped.. :( havent made any progress for 3 days now... not sure exactly what to tweak in GKE. 

In GKE i simply create a cluster. `run kuberctl create -f ... ` with fabric8-console's json. and thats it! all setup, fabric8 console working... but pod console not showing... 

A detailed analysis of the [bugzilla link](https://bugzilla.redhat.com/show_bug.cgi?id=1243317) shows that even eventually a it was fixed as a bug in openshift? not sure this has been done by google guys in kubernetes? 

I basically get the high level picture... the API server is trying to connect to a local ip of a pod but it cannot reach it... but how to correct it? in the the GKE route table? where..... i'm stumped here and im not very used to kubernetes networking... any help will be really appreciated.. thanks

**Comment by michaelukpong** [comment]
@gashcrumb and @rawlingsj  thanks a lot! after days and days! i hollered the GKE guys and it seems to be a bug on their end. see the [topic discussion](https://groups.google.com/forum/#!topic/google-containers/dNJT1TLTwNc)   They have raised this as a [bug/issue](https://github.com/kubernetes/kubernetes/issues/17172)

But in the mean time it means do pod console for me :( 

Notice the diff between the two proxies:
1. Valid on Openshift and Kubernetes
`http://localhost:8001/api/v1/proxy/namespaces/default/pods/nginx-y0iu3:80/`
and
2. Valid only on Openshift at the moment...
`http://localhost:8001/api/v1/namespaces/default/pods/nginx-y0iu3:80/proxy/`

Until they get that fixed any pointers on a workaround? perhaps If I know were that is invoked on fabric8-console i can git pull the source and change it locally and deploy a custom fabric8-console image?

Thanks for any pointers.

**Comment by gashcrumb** [comment]
@michaelukpong bummer!

We build up that URL that's passed over to the java console [here](https://github.com/hawtio/hawtio-kubernetes/blob/master/plugins/kubernetes/ts/kubernetesModel.ts#L644-L645), we do have code that checks to see if we're on openshift or kubernetes [here](https://github.com/hawtio/hawtio-kubernetes/blob/master/plugins/kubernetes/ts/watcher.ts#L67-L87), so what we could do is use the first form of that URL for kubernetes and the second for openshift, let me create an issue in hawtio-kubernetes to track it.

**Comment by michaelukpong** [comment]
cool @gashcrumb lemme check it out..

**Comment by gashcrumb** [comment]
@michaelukpong if I can get docker pull to stop hanging on me I might be able to push a change as a workaround here...

**Comment by michaelukpong** [comment]
@gashcrumb haha.... i'm praying hard

**Comment by gashcrumb** [comment]
@michaelukpong k, pushed that workaround and kicked off a 1.0.73 build of the fabric8-console image -> https://circleci.com/gh/fabric8io/fabric8-console/415 so you should be able to test it once that job is finished, fingers crossed!

**Comment by michaelukpong** [comment]
great... i think its pushed, i'll change the image in the json  to `docker.io/fabric8/fabric8-console:v1.0.73` and (re) deploy fabric8 on GKE... fingers crossed..

**Comment by michaelukpong** [comment]
@gashcrumb false alarm... tag not yet in hub.docker.... i guess i'll wait a bit for it to be fully pushed...

**Comment by michaelukpong** [comment]
@gashcrumb eerm.... can't find v1.0.73 tag on hub.docker.com, circleci says image pushed.... if i use `"image": "docker.io/fabric8/fabric8-console:v1.0.73"` i bet deploy will fail...... should i use `"image": "docker.io/fabric8/fabric8-console:latest"` instead?

**Comment by jimmidyson** [comment]
Try the v1.0.73 tag - I've seen docker hub not display tags properly for a while. Guess there's some cache.

**Comment by jimmidyson** [comment]
Hang on it looks like it's only pushed latest tag. Assume it wasn't git tagged yet.

**Comment by michaelukpong** [comment]
yea @jimmidyson on github tag v1.0.73 definitely exists...  on docker not yet... and on docker tags go as high as 2.2.67?  .... not quick clear what tag version to pull from docker.

**Comment by gashcrumb** [comment]
Ah, think actually that build just pushes latest now.

**Comment by gashcrumb** [comment]
pull latest and you should be able to test it.

**Comment by michaelukpong** [comment]
@gashcrumb @jimmidyson  guys i'm pulling my hair here....  the docs in [fabric8 getting started](http://fabric8.io/guide/getStarted/gke.html) point to maven repo  2.2.19.. when i browsed there i noticed there was a 2.2.23.1 also available, and as I expect, since its more recent i decided to go with that version. The kubernetes.json in 2.2.23.1 points to --> `"image":"docker.io/fabric8/fabric8-console:v1.0.29"` and thats what i have been using all along which mostly worked except the issue of viewing pod consoles as documented in this thread...

Now, after @gashcrumb's docker push. I check hub.docker.com and i noticed (my first time checking) that the image version/tags have progressed quite a bit... way ahead of the the "v1.0.29" in the maven repo.  I have tried both `"image":"docker.io/fabric8/fabric8-console:2.2.67"` and `"image":"docker.io/fabric8/fabric8-console:latest"` on GKE NOTHING seems to show... i cant even see the first fabric8-console page... so I cant even get to the point earlier, where the fabric8-console was visible but you couldnt view pod consoles.

```
[Developer] Failed to load https://146.148.35.236/api/v1/proxy/namespaces/default/services/templates/oapi/v1/projects  404
https://146.148.35.236/oapi/v1/namespaces/default/builds Failed to load resource: the server responded with a status of 404 (Not Found)
https://146.148.35.236/oapi/v1/namespaces/default/policybindings Failed to load resource: the server responded with a status of 404 (Not Found)
```

First thing i notice is the use of `oapi` ? 

A bit of digging deep i notice that whereas
`https://146.148.35.236/api/v1/proxy/namespaces/default/services/templates/oapi/v1/projects` isnt valid on GKE, but `https://146.148.35.236/api/v1/proxy/namespaces/default/services/templates/oapi/v1/namespaces/default/templates` is actually valid.... 

I don't even know where to start from :cry:  Now... if i revert back to  `"image":"docker.io/fabric8/fabric8-console:v1.0.29"` things look better, because the main console is showing... except for the pod console issue...   So im just wondering... is this a lost cause? is there hope that fabric8-console will ever work smoothly on GKE?

**Comment by gashcrumb** [comment]
@michaelukpong some of those 404s would be expected, think I can do a bit of work though to cut down on them, right now our client-side code is blindly trying to fetch all the things.  If you can keep that instance running I think I can point my dev env at it which will make it a lot easier to fix up the console.  Will be on IRC in a few...

**Comment by michaelukpong** [comment]
ok @gashcrumb im joining irc....

**Comment by gashcrumb** [comment]
Getting somewhere, here's the java console of a pod for example:
![fabric9-console-gke](https://cloud.githubusercontent.com/assets/351660/11154990/99402962-8a10-11e5-88a6-c4c0c1f72fa5.png)
There's been a few big changes in the console since even this issue was raised, one was a big overhaul of the navigation, unfortunately the root of the app right now needs BuildConfigs to show anything, but that's openshift-specific.  

Need to work through how we can get around this for vanilla kubernetes, for the moment though you can at least get to a namespace by tacking on '/kubernetes/namespace/<namespace name>', then all the links in the tabs should work alright, had to go through and fix up a bunch of them.  

The connect links should work also, however the 'terminal' icons won't work I don't think, should probably remove them for now unless the backend part of that gets pushed upstream, think that's using some openshift specific fix as well.

**Comment by michaelukpong** [comment]
Hi @gashcrumb i confess i left this for a while to concentrate on my core code.. so... good progress... :) which docker image version should i use in the kubernetes.json to get to at least this point you are getting to?

**Comment by michaelukpong** [comment]
@gashcrumb :latest worked... so yeah.. some progress :smile:
